NitroxServer/GameLogic/StoryManager.cs

```diff
@@ -0,0 +1,268 @@
+
+    public Timer ResyncTimer;
+    /// <summary>
+    /// Time in seconds between each resync packet sending.
+    /// </summary>
+    private const int RESYNC_INTERVAL = 60;
+
+    public StoryManager(PlayerManager playerManager, PDAStateData pdaStateData, StoryGoalData storyGoalData, string seed, double elapsedSeconds, double? auroraExplosionTime, double? auroraWarningTime)
+    {
+        this.playerManager = playerManager;
+        this.pdaStateData = pdaStateData;
+        this.storyGoalData = storyGoalData;
+        this.seed = seed;
+        // Default time in Base SN is 480s
+        elapsedTimeOutsideStopWatchMs = elapsedSeconds == 0 ? TimeSpan.FromSeconds(480).TotalMilliseconds : elapsedSeconds * 1000;
+        AuroraCountdownTimeMs = auroraExplosionTime ?? GenerateDeterministicAuroraTime(seed);
+        AuroraWarningTimeMs = auroraWarningTime ?? ElapsedTimeMs;
+        SetupResyncInterval();
+    }
+
+    /// <summary>
+    /// Creates a timer that periodically sends resync packets to players
+    /// </summary>
+    public void SetupResyncInterval()
+    {
+        ResyncTimer = new()
+        {
+            Interval = TimeSpan.FromSeconds(RESYNC_INTERVAL).TotalMilliseconds,
+            AutoReset = true
+        };
+        ResyncTimer.Elapsed += delegate
+        {
+            playerManager.SendPacketToAllPlayers(new AuroraAndTimeUpdate(GetInitialTimeData(), false));
+        };
+    }
+
+    /// <summary>
+    /// Tells the players to start Aurora's explosion event
+    /// </summary>
+    /// <param name="countdown">Wether we should make Aurora explode instantly or after a short countdown</param>
+    public void ExplodeAurora(bool countdown)
+    {
+        // Calculations from CrashedShipExploder.OnConsoleCommand_countdownship()
+        // We add 3 seconds to the cooldown so that players have enough time to receive the packet and process it
+        AuroraCountdownTimeMs = ElapsedTimeMs + 3000;
+        AuroraWarningTimeMs = AuroraCountdownTimeMs;
+
+        if (countdown)
+        {
+            Log.Info("Aurora's explosion countdown will start in 3 seconds");
+        }
+        else
+        {
+            // Calculations from CrashedShipExploder.OnConsoleCommand_explodeship()
+            AuroraCountdownTimeMs -= 25000;
+            AuroraWarningTimeMs -= 1000;
+            Log.Info("Aurora's explosion initiated");
+        }
+
+        playerManager.SendPacketToAllPlayers(new AuroraAndTimeUpdate(GetInitialTimeData(), false));
+    }
+
+    /// <summary>
+    /// Tells the players to start Aurora's restoration event
+    /// </summary>
+    public void RestoreAurora()
+    {
+        AuroraWarningTimeMs = ElapsedTimeMs;
+        AuroraCountdownTimeMs = GenerateDeterministicAuroraTime(seed);
+
+        // We need to clear these entries from PdaLog and CompletedGoals to make sure that the client, when reconnecting, doesn't have false information
```

Comment:
How is a reconnecting client relevant here?

Reply:
When restoring aurora, we want to delete the aurora warnings that already happened from the completed goals list so that they'll work correctly when called another time. But the deletion of these goals on client-side must be replicated here on server-side. If the goals aren't deleted on server-side, a user that will disconnect and reconnect back will get the unmodified (modifications that happened on the client when receiving this packet) list of completed goals and therefore, when the events will be triggered, they won't work correctly because they'll be registered as if they already happened.